### PR TITLE
[DPC-4220] Add max failed attempts to invite verification

### PR DESCRIPTION
**Pull request by ashley-weaver**
## 🎫 Ticket

https://jira.cms.gov/browse/DPC-4220

## 🧪 Validation
Invalid code error
image

Max tries exceeded error
image

### CODE REVIEW

File: dpc-portal/app/controllers/invitations_controller.rb [hunk 1]
```diff
@@ -39,15 +39,25 @@ def code
...
-      return render(Page::Invitations::OtpComponent.new(@organization, @invitation), status: :bad_request)
-    end
+    return add_failed_attempt unless params[:verification_code] == @invitation.verification_code
 
+    @invitation.reset_attempts
     session["invitation_status_#{@invitation.id}"] = 'code_verified'
     render(Page::Invitations::InvitationLoginComponent.new(@invitation))
   end
 
+  def add_failed_attempt
```
Comment (jdettmannnava) [inline dpc-portal/app/controllers/invitations_controller.rb:49]: I would call this 'handle_failed_attempt'

File: dpc-portal/app/controllers/invitations_controller.rb [hunk 2]
```diff
@@ -39,15 +39,25 @@ def code
...
+    attempts_remaining = @invitation.update_attempts
+    if attempts_remaining.zero?
+      return render(Page::Invitations::BadInvitationComponent.new(@invitation, 'max_tries_exceeded'),
+                    status: :forbidden)
+    end
+
+    @invitation.errors.add(:verification_code, :bad_code,
+                           message: "Incorrect invite code. You have #{attempts_remaining} remaining attempts.")
```
Comment (jdettmannnava) [inline dpc-portal/app/controllers/invitations_controller.rb:57]: The message should probably be a flash alert.

File: dpc-portal/app/models/invitation.rb [hunk 3]
```diff
@@ -19,12 +19,24 @@ class Invitation < ApplicationRecord
...
+  MAX_ATTEMPTS = 5
 
   def phone_raw=(nbr)
     @phone_raw = nbr
     self.invited_phone = @phone_raw.tr('^0-9', '')
   end
 
+  def update_attempts
+    update(failed_attempts: failed_attempts + 1) unless failed_attempts == MAX_ATTEMPTS
+    save
```
Comment (jdettmannnava) [inline dpc-portal/app/models/invitation.rb:31]: Unnecessary; "update" saves

File: dpc-portal/app/models/invitation.rb [hunk 4]
```diff
@@ -19,12 +19,24 @@ class Invitation < ApplicationRecord
...
+    MAX_ATTEMPTS - failed_attempts
+  end
+
+  def reset_attempts
+    update(failed_attempts: 0)
+    save
```
Comment (jdettmannnava) [inline dpc-portal/app/models/invitation.rb:37]: Unnecessary

File: dpc-portal/app/models/invitation.rb [hunk 5]
```diff
@@ -19,12 +19,24 @@ class Invitation < ApplicationRecord
...
   STEPS = ['Sign in or create a Login.gov account', 'Confirm your identity', 'Confirm organization registration',
            'Finished'].freeze
+  MAX_ATTEMPTS = 5
 
   def phone_raw=(nbr)
     @phone_raw = nbr
     self.invited_phone = @phone_raw.tr('^0-9', '')
   end
 
+  def update_attempts
```
Comment (jdettmannnava) [inline dpc-portal/app/models/invitation.rb:29]: maybe increment_failed_attempts

**Review by jdettmannnava — COMMENTED** [review]
We'll also need to add max_tries_exceeded to Invitation#uncacceptable_reason

File: dpc-portal/app/controllers/invitations_controller.rb (hunk 2 above)
Comment (ashley-weaver) [reply dpc-portal/app/controllers/invitations_controller.rb:57]: instead of adding to the errors, or as well as?

Comment (jdettmannnava) [reply dpc-portal/app/controllers/invitations_controller.rb:57]: I think "Incorrect invite code" should stay here. and "Incorrect invite code. You have..." should be in flash

File: dpc-portal/app/controllers/invitations_controller.rb [hunk 6]
```diff
@@ -39,22 +39,23 @@ def code
...
     @invitation.reset_attempts
     session["invitation_status_#{@invitation.id}"] = 'code_verified'
     render(Page::Invitations::InvitationLoginComponent.new(@invitation))
   end
 
-  def add_failed_attempt
-    attempts_remaining = @invitation.update_attempts
+  def handle_failed_attempt
+    @invitation.increment_failed_attempts
+    attempts_remaining = @invitation.reload.attempts_remaining
```
Comment (jdettmannnava) [inline dpc-portal/app/controllers/invitations_controller.rb:51]: does this need a reload?

File: dpc-portal/app/models/invitation.rb [hunk 7]
```diff
@@ -93,6 +94,7 @@ def unacceptable_reason # rubocop:disable Metrics/CyclomaticComplexity,Metrics/P
     return 'invalid' if cancelled?
     return 'accepted' if accepted?
     return 'ao_renewed' if renewed? && authorized_official?
+    return 'max_tries_exceeded' if attempts_remaining.zero?
```
Comment (jdettmannnava) [inline dpc-portal/app/models/invitation.rb:97]: This needs a test

File: dpc-portal/config/locales/en.yml [hunk 8]
```diff
@@ -14,7 +14,8 @@ en:
     ao_med_sanctions_text: You are listed as an excluded individual by the Office of Inspector General (OIG), which prohibits you from acting as this organization’s Authorized Official. For more information, contact the <a href="https://oig.hhs.gov/about-oig/contact-us/#exclusions">OIG Exclusions Branch</a>.
     org_med_sanctions_status: Your organization is in the Medicare Exclusions Database.
     org_med_sanctions_text: Your organization is listed as an excluded entity by the Office of Inspector General (OIG), which prohibits it from accessing beneficiary claims data. For more information, contact the <a href="https://oig.hhs.gov/about-oig/contact-us/#exclusions">OIG Exclusions Branch</a>.
-
+    max_tries_exceeded_status: Your access is locked due to too many attempts.
+    max_tries_exceeded_text: You have made too many attempts. Please request a new invite from the Authorized Official.
```
Comment (jdettmannnava) [inline dpc-portal/config/locales/en.yml:18]: I just noticed; Eric wants the AO's name interpolated here

Comment (ashley-weaver) [reply dpc-portal/config/locales/en.yml:18]: would it have to be handled outside of this file then?

Comment (jdettmannnava) [reply dpc-portal/config/locales/en.yml:18]: No, rails supports locale interpolation: https://guides.rubyonrails.org/i18n.html#passing-variables-to-translations
We used to have examples in this file, but no more!

File: dpc-portal/app/components/page/invitations/bad_invitation_component.rb [hunk 9]
```diff
@@ -8,6 +8,7 @@ def initialize(invitation, reason)
         super
         @invitation = invitation
         @org_name = invitation&.provider_organization&.name
+        @ao_display_name = "#{invitation&.invited_given_name} #{invitation&.invited_family_name}"
```
Comment (jdettmannnava) [inline dpc-portal/app/components/page/invitations/bad_invitation_component.rb:11]: This is the cd...
invitation invited_by is the AO

File: dpc-portal/app/components/page/invitations/bad_invitation_component.html.erb [hunk 10]
```diff
@@ -2,12 +2,12 @@
...
   <div>
     <p>
-      <%=raw t(key=@text, org_name: @org_name) %>
+      <%=raw t(key=@text, org_name: @org_name, ao_full_name: @ao_full_name) %>
     </p>
   </div>
   <% 'have to put the case statement here, as do not have route helper in ViewComponent'
     case(@reason)
-      when :ao_expired %>
+      when :ao_expired, :max_tries_exceeded %>
```
Comment (jdettmannnava) [inline dpc-portal/app/components/page/invitations/bad_invitation_component.html.erb:10]: Why?

Comment (ashley-weaver) [reply dpc-portal/app/components/page/invitations/bad_invitation_component.html.erb:10]: oh my bad, this is the AO request link, nvm

**Review by jdettmannnava — APPROVED** [review]
LGTM

**Pull request merged**